{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"cross_tools.ipynb","provenance":[],"collapsed_sections":[],"authorship_tag":"ABX9TyM+eT7v7xi8uIfBwuixPnRk"},"kernelspec":{"name":"python3","display_name":"Python 3"}},"cells":[{"cell_type":"code","metadata":{"id":"b5Fa2lSalDHB","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1615266534140,"user_tz":420,"elapsed":882,"user":{"displayName":"Ammon Knaupp","photoUrl":"","userId":"01156806751079449669"}},"outputId":"861dae9d-cbad-455b-ca96-e11c5ad4f33c"},"source":["class Assortment:\r\n","  def __init__(self, a1, a2, **kwargs):\r\n","    self.allele1=a1\r\n","    self.allele2=a2\r\n","    self.gene_name='(default)'\r\n","    self.assortment=a1+a2\r\n","    #self.dominant_phenotype='(default)'\r\n","    #self.recessive_phenotype='(default)'\r\n","    # if provided by user, assign dominance\r\n","    if 'allele1_is_dominant' in kwargs: self.allele1_is_dominant=kwargs['allele1_is_dominant']\r\n","    else:\r\n","      if self.allele1.islower(): self.allele1_is_dominant=False # otherwise, if allele is all lowercase, assign dominance as recessive\r\n","      else: self.allele1_is_dominant=True\r\n","    if 'allele2_is_dominant' in kwargs: self.allele2_is_dominant=kwargs['allele2_is_dominant']\r\n","    else:\r\n","      if self.allele2.islower(): self.allele2_is_dominant=False # otherwise, if allele is all lowercase, assign dominance as recessive\r\n","      else: self.allele2_is_dominant=True\r\n","    # initialize zygosity\r\n","    if self.allele1_is_dominant and self.allele2_is_dominant: self.zygosity='Homozygous Dominant'\r\n","    elif not self.allele1_is_dominant and not self.allele2_is_dominant: self.zygosity='Homozygous Recessive'\r\n","    else: self.zygosity='Heterozygous'\r\n","    # if provided by user, assign phenotypes\r\n","    if 'dominant_phenotype' in kwargs: \r\n","      self.dominant_phenotype=kwargs['dominant_phenotype']\r\n","    if 'recessive_phenotype' in kwargs: \r\n","      self.recessive_phenotype=kwargs['recessive_phenotype']\r\n","  def _print(self):\r\n","    output_string=self.gene_name+' '+self.allele1\r\n","    if self.allele1_is_dominant:\r\n","      output_string+=' (dom), '\r\n","    else: output_string+=' (rec), '\r\n","    output_string+=self.allele2\r\n","    if self.allele2_is_dominant:\r\n","      output_string+=' (dom)'\r\n","    else: output_string+=' (rec)'\r\n","    return output_string\r\n","  def __eq__(self, other): # declares equivalence between two genes with the same alleles in any order.\r\n","    return ( self.allele1+self.allele2 == other.allele1+other.allele2 or self.allele1+self.allele2 == other.allele2+other.allele1 )\r\n","\r\n","class Three_factor_linkage:\r\n","  def __init__(self, assortment1, assortment2, assortment3):\r\n","    self.gene1 = assortment1\r\n","    self.gene2 = assortment2\r\n","    self.gene3 = assortment3\r\n","    self.non_recomb=[[self.gene1.allele1, self.gene2.allele1, self.gene3.allele1],[self.gene1.allele2, self.gene2.allele2, self.gene3.allele2]]\r\n","    self.single_crossover=[[self.gene1.allele1, self.gene2.allele1, self.gene3.allele2],\r\n","                           [self.gene1.allele1, self.gene2.allele2, self.gene3.allele2],\r\n","                           [self.gene1.allele2, self.gene2.allele2, self.gene3.allele1],\r\n","                           [self.gene1.allele2, self.gene2.allele1, self.gene3.allele1]]\r\n","    self.double_crossover=[[self.gene1.allele1, self.gene2.allele2, self.gene3.allele1],[self.gene1.allele2, self.gene2.allele1, self.gene3.allele2]]\r\n","    self.gametes=self.non_recomb + self.single_crossover + self.double_crossover\r\n","    self.genotype='Gene 1: {}\\nGene 2: {}\\nGene 3: {}\\n'.format(self.gene1.assortment,self.gene2.assortment,self.gene3.assortment)\r\n","    self.chr1=self.gene1.allele1 + self.gene2.allele1 + self.gene3.allele1\r\n","    self.chr2=self.gene1.allele2 + self.gene2.allele2 + self.gene3.allele2\r\n","  def print_gametes(self):\r\n","    a='non-recomb1:\\t\\t{}-{}-{}'.format(self.gene1.allele1, self.gene2.allele1, self.gene3.allele1)\r\n","    b='non-recomb2:\\t\\t{}-{}-{}'.format(self.gene1.allele2, self.gene2.allele2, self.gene3.allele2)\r\n","    c='single cross-over1:\\t{}-{}-{}'.format(self.gene1.allele1, self.gene2.allele1, self.gene3.allele2)\r\n","    d='single cross-over2:\\t{}-{}-{}'.format(self.gene1.allele1, self.gene2.allele2, self.gene3.allele2)\r\n","    e='single cross-over3:\\t{}-{}-{}'.format(self.gene1.allele2, self.gene2.allele2, self.gene3.allele1)\r\n","    f='single cross-over4:\\t{}-{}-{}'.format(self.gene1.allele2, self.gene2.allele1, self.gene3.allele1)\r\n","    g='double cross-over1:\\t{}-{}-{}'.format(self.gene1.allele1, self.gene2.allele2, self.gene3.allele1)\r\n","    h='double cross-over2:\\t{}-{}-{}'.format(self.gene1.allele2, self.gene2.allele1, self.gene3.allele2)\r\n","    return '{}\\n{}\\n{}\\n{}\\n{}\\n{}\\n{}\\n{}\\n'.format(a,b,c,d,e,f,g,h)\r\n","  def reorder(self,order):\r\n","    tmp1=self.gene1\r\n","    tmp2=self.gene2\r\n","    tmp3=self.gene3\r\n","    if order[0] == 2: self.gene2=tmp1\r\n","    if order[0] == 3: self.gene3=tmp1\r\n","    if order[1] == 1: self.gene1=tmp2\r\n","    if order[1] == 3: self.gene3=tmp2\r\n","    if order[2] == 1: self.gene1=tmp3\r\n","    if order[2] == 2: self.gene2=tmp3\r\n","\r\n","class Two_factor_unlinked:\r\n","  def __init__(self, assortment1, assortment2):\r\n","    self.gene1 = assortment1\r\n","    self.gene2 = assortment2\r\n","    self.non_recombs=[[self.gene1.allele1, self.gene2.allele1],[self.gene1.allele2, self.gene2.allele2]]\r\n","    self.recombinants=[[self.gene1.allele1, self.gene2.allele2],[self.gene1.allele2, self.gene2.allele1]]\r\n","    self.gametes=self.non_recombs + self.recombinants\r\n","    self.genotype='{}, {}'.format(self.gene1.assortment, self.gene2.assortment)\r\n","    self.chr1=self.gene1.allele1 + self.gene2.allele1\r\n","    self.chr2=self.gene1.allele2 + self.gene2.allele2\r\n","  def print_gametes(self):\r\n","    a='non-recomb1:\\t{}-{}'.format(self.gene1.allele1, self.gene2.allele1)\r\n","    b='non-recomb2:\\t{}-{}'.format(self.gene1.allele2, self.gene2.allele2)\r\n","    c='recombinant1:\\t{}-{}'.format(self.gene1.allele1, self.gene2.allele2)\r\n","    d='recombinant2:\\t{}-{}'.format(self.gene1.allele2, self.gene2.allele1)\r\n","    return '{}\\n{}\\n{}\\n{}\\n'.format(a,b,c,d)\r\n","  def __eq__(self, other):\r\n","    return (self.gene1 == other.gene1 and self.gene2 == other.gene2)\r\n","  def cross(self, mate):\r\n","    zygotes=[]\r\n","    products=[]\r\n","    for i in range(0,4):\r\n","      for j in range(0,4):\r\n","        products.append([[self.gametes[i][0], mate.gametes[j][0]], [self.gametes[i][1], mate.gametes[j][1]]])\r\n","    for product in products: ## create Two_factor_unlinked objects\r\n","      new_gene1 = Assortment(product[0][0], product[0][1], dominant_phenotype=self.gene1.dominant_phenotype, recessive_phenotype=self.gene1.recessive_phenotype)\r\n","      new_gene2 = Assortment(product[1][0], product[1][1], dominant_phenotype=self.gene2.dominant_phenotype, recessive_phenotype=self.gene2.recessive_phenotype)\r\n","      zygote = Two_factor_unlinked(new_gene1, new_gene2)\r\n","      zygotes.append(zygote)\r\n","    return Two_factor_cross_progeny(zygotes)\r\n","\r\n","class Two_factor_cross_progeny:\r\n","  def __init__(self, zygotes):\r\n","    self.zygotes = zygotes\r\n","    self.genotype_counts = []\r\n","    self.genotypes = []\r\n","    for zygote in zygotes: # populate genotypes with only unique zygotes and find the genotype ratio\r\n","      alreadyThere = False\r\n","      for i,genotype in enumerate(self.genotypes): # check genotypes to see if current zygote should be added\r\n","        if zygote == genotype:\r\n","          alreadyThere = True\r\n","      if alreadyThere: # if this zygote is already in genotypes\r\n","        self.genotype_counts[i]+=1 # increment count of this zygote\r\n","      else:\r\n","        self.genotypes.append(zygote) # this zygote isn't represented in genotypes. Add it and initialize its count in the ratio to 1.\r\n","        self.genotype_counts.append(1)\r\n","    self.zygosities = {}\r\n","    for zygote in zygotes:\r\n","      zygosity = zygote.gene1.zygosity + ', ' + zygote.gene2.zygosity\r\n","      if zygosity in self.zygosities.keys(): # if zygosity is already in dictionary\r\n","        self.zygosities[zygosity]+=1 # increment count of zygosity\r\n","      else:\r\n","        self.zygosities[zygosity]=1\r\n","    self.phenotypic_ratio = {}\r\n","    for zygote in zygotes:\r\n","      phenotype = ''\r\n","      if zygote.gene1.zygosity == 'Homozygous Recessive':\r\n","        phenotype += zygote.gene1.recessive_phenotype + ', '\r\n","      else: phenotype = zygote.gene1.dominant_phenotype + ', '\r\n","      if zygote.gene2.zygosity == 'Homozygous Recessive':\r\n","        phenotype += zygote.gene2.recessive_phenotype\r\n","      else: phenotype += zygote.gene2.dominant_phenotype\r\n","      if phenotype in self.phenotypic_ratio.keys(): # if phenotype is already in dictionary\r\n","        self.phenotypic_ratio[phenotype]+=1 # increment count of phenotype\r\n","      else:\r\n","        self.phenotypic_ratio[phenotype]=1\r\n","  def print_ratio(self): # genotypic ratio\r\n","    output_string=''\r\n","    for i in range(len(self.genotype_counts)):\r\n","      output_string+=self.genotypes[i].genotype + ': ' + str(self.genotype_counts[i]) + '\\n'\r\n","    return output_string\r\n","\r\n","class Three_factor_cross_progeny:\r\n","  def __init__(self, zygotes):\r\n","    self.zygotes = zygotes\r\n","    self.ratio = []\r\n","    self.genotypes = []\r\n","    for zygote in zygotes: # populate genotypes with only unique zygotes and find the genotype ratio\r\n","      alreadyThere = False\r\n","      for i,genotype in enumerate(self.genotypes): # check genotypes to see if current zygote should be added\r\n","        if zygote == genotype:\r\n","          alreadyThere = True\r\n","      if alreadyThere: # if this zygote is already in genotypes\r\n","        self.ratio[i]+=1 # increment count of this zygote in the ratio\r\n","      else:\r\n","        self.genotypes.append(zygote) # this zygote isn't represented in genotypes. Add it and initialize its count in the ratio to 1.\r\n","        self.ratio.append(1)\r\n"],"execution_count":25,"outputs":[{"output_type":"stream","text":["\n","FRUITFLY TESTCROSS EXAMPLE 2:\n","Cross between red eyed, wild type winged parent and a white eyed, vestigial winged parent.\n","This time, the second parent is heterozygous for both traits. The first parent remains homozygous recessive for both.\n","\n","Progeny display only the dominant phenotype, but this time, four genotypes exist:\n","{'Red Eyes, WT wings': 4, 'White Eyes, Vestigial wings': 4, 'Red Eyes, Vestigial wings': 4, 'White Eyes, WT wings': 4}\n","Genotype  1 :  wW, bB . Zygosity: Heterozygous , Heterozygous\n","Genotype  2 :  ww, bb . Zygosity: Homozygous Recessive , Homozygous Recessive\n","Genotype  3 :  wW, bb . Zygosity: Heterozygous , Homozygous Recessive\n","Genotype  4 :  ww, bB . Zygosity: Homozygous Recessive , Heterozygous\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"j4YvvNYFghFx","executionInfo":{"status":"ok","timestamp":1614742045934,"user_tz":420,"elapsed":665,"user":{"displayName":"Ammon Knaupp","photoUrl":"","userId":"01156806751079449669"}},"outputId":"1aec2c05-139e-4a8d-f427-5f8d66b06291"},"source":[" # bifid phenotype: wings shortened, fused veins, oriented outward. Caused by recessive allele.\r\n"," # white phenotype: eyes white. Caused by recessive allele 'w'.\r\n"," # rudimentary phenotype: wings shortened and blistered. Caused by recessive allele 'r'.\r\n"," # bi, w, r\r\n","\r\n","white_bifid_female = Three_factor_linkage(Assortment('b','b'), Assortment('w','w'), Assortment('WT','WT')) # p1\r\n","rudimentary_male = Three_factor_linkage(Assortment('WT','WT'), Assortment('WT','WT'), Assortment('r','r')) # p2\r\n","\r\n","# All F1 females were wild type and all F1 males were white, bifid.\r\n","f1_females = Three_factor_cross_progeny([white_bifid_female])\r\n","f2_males = Three_factor_cross_progeny([])\r\n","\r\n","# They allowed the F1 females and males to mate with one another to produce an F2 generation.\r\n","# They noticed that the rudimentary phenotype causes partial loss of viability (i.e. some of the rudimentary flies died during early developmental stages and thus could not be counted).\r\n","# Nonetheless, knowing the phenotypes and genotypes of the original parents, they were able to correctly map the genes.\r\n","\r\n","# The following lists the phenotypes of the male flies only in the F2 generation. "],"execution_count":null,"outputs":[{"output_type":"stream","text":["d\n"],"name":"stdout"}]}]}