{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"ploidy.ipynb","provenance":[],"mount_file_id":"1j7oMQHnXnGhtBZ7wmD00hqPze2XGoQC6","authorship_tag":"ABX9TyO6J1Von4IRp/F2gcsFMHiP"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","metadata":{"id":"N22Qtpz08GSy"},"source":["# Ploidy Python Library"]},{"cell_type":"code","metadata":{"id":"tZgO4GpaNLLH","executionInfo":{"status":"ok","timestamp":1616452317503,"user_tz":360,"elapsed":411,"user":{"displayName":"Ammon Knaupp","photoUrl":"","userId":"01156806751079449669"}}},"source":["# n, (the 'haploid number') is the number of chromosomes in a gamete. equal to half of total chromosome count. Hence, 2n = total chromosomes.\n","# If the organism undergoes meiosis and non-disjunction occurs, n is doubled.\n","# c_n (the coefficient of n) corresponds to whether the organism is a gamete or not. once an organism undergoes meiosis, c_n = 1. After conception and prior to meiosis, c_n = 2.\n","# c_n has a value of 1 if the organism is a gamete (during and after Meiosis II) and 2 in any other case.\n","# the value of x corresponds to the number of chromosomes which form a complete set (basically, it is the number of chromosomes in a genome).\n","# two or more chromosomes are said to be homeologous if they are very similar, yet cannot synapse during meiosis I.\n","# the value of c_x corresponds to the number of chromosome sets that are homeologous. indicated by type of ploidy (tri-, tetra-, quadra-, penta-, hexa-)\n","# the value of c_x is divided by two during meiosis, assuming non-disjunction doesn't occur. if non-disjunction occurs, then c_x doesn't change.\n","\n","# a genome has an x number, a species has an n number and c_x.\n","\n","class Polyploidic_Assortment:\n","  def __init__(self, *args, **kwargs):\n","    self.assortment = args\n","    self.dominant_allele = 'default!'\n","    self.gametes=[[self.assortment[0],self.assortment[1]],[self.assortment[2],self.assortment[3]],[self.assortment[0],self.assortment[2]],[self.assortment[1],self.assortment[3]],[self.assortment[0],self.assortment[3]],[self.assortment[1],self.assortment[2]]]\n","    # determine dominant allele\n","    if 'dominant_allele' in kwargs: self.dominant_allele=kwargs['dominant_allele']\n","    else:\n","      for allele in self.assortment:\n","        if allele[0].isupper(): self.dominant_allele = allele # if all alleles are lowercase ???\n","      assert (self.dominant_allele != 'default!'),'Failed to create Polyploidic_Allele object. No dominant_allele provided, no alleles were uppercase.'\n","      print('Dominant allele \"{}\" selected due to uppercase. Alternatively, pass dominant_allele as keyword argument.'.format(self.dominant_allele))\n","    # count dominant alleles, initialize plexy\n","    num_dominant_alleles = self.assortment.count(self.dominant_allele)\n","    plexies = ['nulli','sim','du','tri','quadri']\n","    self.plexy = plexies[num_dominant_alleles] + 'plex'\n","    # if provided by user, assign phenotypes\n","    if 'dominant_phenotype' in kwargs: \n","      self.dominant_phenotype=kwargs['dominant_phenotype']\n","    if 'recessive_phenotype' in kwargs: \n","      self.recessive_phenotype=kwargs['recessive_phenotype']\n","  def cross(self, mate):\n","    products=[]\n","    for i in range(0,6):\n","      for j in range(0,6):\n","        products.append(Polyploidic_Assortment(self.gametes[i][0], self.gametes[i][1], mate.gametes[j][0], mate.gametes[j][1], dominant_allele=self.dominant_allele).plexy)\n","    plexies = []\n","    for plexy in products:\n","      if plexy not in plexies: plexies.append(plexy)\n","    ratio = {'nulliplex':0,'simplex':0,'duplex':0,'triplex':0,'quadriplex':0}\n","    for plexy in plexies:\n","      ratio[plexy] = products.count(plexy)\n","    return ratio\n","\n","class Monoploid_Genome:\n","  def __init__(self, name, x=0):\n","    self.name = name\n","    self.x = x\n","\n","class Species: # ASSUME for now that x must be the same for all genomes\n","  def __init__(self, *args, **kwargs):\n","    self.genome = []\n","    self.prefixes = ['ha','di','tri','tetra','penta','hexa']\n","    self.ploidy = self.prefixes[len(args)-1] + 'ploid' # initialize, modify later if necessary\n","    self.total_chromosomes=0\n","    if len(args) == 1: # OPTION 1: parameters are (one) Monoploid_Genome, total_chromosomes, and x (unless Monoploid_Genome object x != 0)\n","      assert ('total_chromosomes' in kwargs),'Species object cannot be created. Species __init__ assumes if only one genome is provided that total_chromosomes and x will be provided.'\n","      self.total_chromosomes = kwargs['total_chromosomes']\n","      if args[0].x == 0:\n","        assert ('x' in kwargs),'Species object cannot be created. x must be provided, either to Species __init__ or Monoploid_Genome __init__.'\n","        self.x = kwargs['x'] # x passed as parameter\n","      else: self.x = args[0].x # x was passed as parameter to Monoploid_Genome object, copy its value\n","      assert (self.total_chromosomes % self.x == 0),'Species object cannot be created. Total chromosomes must be a multiple of x.'\n","      self.c_x = int(self.total_chromosomes / self.x)\n","      self.ploidy = self.prefixes[self.c_x-1] + 'ploid'\n","    elif len(args) > 1: # OPTION 2: parameters are multiple Monoploid_Genome objects\n","      assert (args[0].x == 0),'Species object cannot be created. Not all Monoploid_Genome objects have an x value, therefore total_chromosomes must be passed to Species __init__.'\n","      self.x = args[0].x # arbitrarily define x to match the number of chromosomes in the first genome.\n","      all_x_provided=True\n","      for monoploid_genome in args:\n","        if (monoploid_genome.x == 0): all_x_provided=False\n","          #'\"{}\" genome has more or less chromosomes than the first genome belonging to this Species.'.format(monoploid_genome.name)\n","        self.genome.append(monoploid_genome)\n","      if not all_x_provided:\n","        assert ('total_chromosomes' in kwargs),'Species object cannot be created. Not all Monoploid_Genome objects have an x value, therefore total_chromosomes must be passed to Species __init__.'\n","      if 'total_chromosomes' in kwargs: self.total_chromosomes = kwargs['total_chromosomes']\n","      else: self.total_chromosomes = self.c_x * self.x\n","      if self.total_chromosomes % 2 != 0: print('total_chromosomes is odd. value of n (unless provided as parameter) will be calculated truncating decimals from value of (total_chromosomes / 2).')\n","    # initialize derivative variables\n","    if 'n' in kwargs: self.n = kwargs['n'] # haploid number\n","    else: self.n = self.total_chromosomes / 2\n","    self.c_n = 2 # coefficient belonging to n\n","    if self.ploidy == 'haploid': self.c_n = 1\n","    # initialize gametes\n","    self.gamete='Gamete: '\n","    self.gametesReduced = True # whether meiosis I disjunction occurs\n","    for monoploid_genome in self.genome:\n","      if self.gametesReduced: self.gamete += monoploid_genome.name\n","      else: self.gamete += (monoploid_genome.name *2)\n","    self.unreduced_gamete = 'Unreduced gamete: '\n","    for monoploid_genome in self.genome:\n","      self.unreduced_gamete += (monoploid_genome.name * 2)\n","    # make sure that all values are consistent with 2n == c_x*x == total_chromosomes. Always true in the case that none of the values are provided by user.\n","    assert ((self.c_x * self.x == self.total_chromosomes and self.x == 0) or 2 * self.n == self.total_chromosomes),'Species __init__: parameter(s) provided are inconsistent with expression \"2n == (coefficient)*x == total_chromosomes\".'\n","  def cross(self, other):\n","    return 'blah'\n","  def toString(self):\n","    return '{}, {}n = {}x = {}. (x={}, n={})'.format(self.ploidy, self.c_n, self.c_x, self.total_chromosomes, self.x, self.n)"],"execution_count":5,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"G2Z8Wih34M0T"},"source":["# Homework 7"]},{"cell_type":"markdown","metadata":{"id":"Y3B8aFuHlmmV"},"source":["### Question 1\n","Cotton was cultivated for millennia as a source of fiber in the Old World. The two most commonly cultivated species were Gossypium herbacium and G. arboreum. Both species have 2n = 26 chromosomes. They differ from each other by a reciprocal translocation. Cotton was also cultivated by ancient civilizations in the New World. The two cultivated species were G. hirsutum and G. barbadense, each with 2n = 52 chromosomes. A total of 26 wild species with 2n = 26 chromosomes have been identified in Africa, South and Central America, Australia, and the Middle East. Two wild species with 2n = 52 chromosomes have been identified, one with a limited range in northern Brazil and a second in Hawaii.  What is the ploidy of the New World cotton cultivated cotton species G. hirsutum and G. barbadense? "]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"8bBO03j1ll5Q","executionInfo":{"status":"ok","timestamp":1616452614713,"user_tz":360,"elapsed":158,"user":{"displayName":"Ammon Knaupp","photoUrl":"","userId":"01156806751079449669"}},"outputId":"05117aca-032a-4d14-ae2b-f20d81f16e91"},"source":["# Reorganize:\n","'''\n","old_cotton = Species(Monoploid_Genome(name='H'), total_chromosomes=26)\n","print('Old world cotton:', old_cotton.toString())\n","\n","new_cotton = Species(Monoploid_Genome(name='H'), total_chromosomes=52)\n","print('New world cotton:', G_hirsutum.toString())\n","'''\n","# Solve:\n","# What is x?\n","\n","# Assumptions:\n","# x is the same for all cottons\n","# all varieties in question have an even ploidy number\n","\n","old_cotton = Species(Monoploid_Genome(name='H'), total_chromosomes=26, x=13)\n","print('Old world cotton:', old_cotton.toString())\n","\n","new_cotton = Species(Monoploid_Genome(name='H'), total_chromosomes=52, x=13)\n","print('New world cotton:', G_hirsutum.toString())\n","\n","# Answer: Tetraploid.\n","# Reasoning: New world varieties with 2n = 52 chromosomes are autotetraploids derived from old world varieties."],"execution_count":14,"outputs":[{"output_type":"stream","text":["Old world cotton: diploid, 2n = 2x = 26. (x=13, n=13.0)\n","New world cotton: tetraploid, 2n = 4x = 52. (x=13, n=26.0)\n"],"name":"stdout"}]},{"cell_type":"markdown","metadata":{"id":"T3_zk6d442xy"},"source":["## Context for questions 2 and 3\n","In alfalfa, an autotetraploid, purple flowers are due to an allele (C1) that shows complete dominance to the recessive c1 allele. White flowers appear when the plants are nulliplex (c1c1c1c1) whereas purple flowers will appear if at least one of the alleles is C1.  A purple-flowered plant is crossed to a white-flowered plant. Half of the F1 progeny are purple-flowered and half are white-flowered.\n","\n"]},{"cell_type":"markdown","metadata":{"id":"eUoANAgD5Pvd"},"source":["### Question 2\n","What is the genotype of the purple-flowered parent?"]},{"cell_type":"code","metadata":{"id":"ajs3RYA95W54","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1616452317692,"user_tz":360,"elapsed":581,"user":{"displayName":"Ammon Knaupp","photoUrl":"","userId":"01156806751079449669"}},"outputId":"3903e236-1f7c-4a0c-b3b0-0f2a49718084"},"source":["simplex_purple = Polyploidic_Assortment('C1','c1','c1','c1', dominant_allele='C1')\n","white = Polyploidic_Assortment('c1','c1','c1','c1', dominant_allele='C1')\n","\n","# F1 progeny\n","# half are white = Polyploidic_Assortment('c1','c1','c1','c1', dominant_allele='C1')\n","# other half are purple, meaning at least one C1 allele each\n","\n","# Assumptions:\n","# Purple parent isn't C1 x 4\n","print(simplex_purple.cross(white))\n","\n","# Answer: in order to achieve 1:1 ratio from a cross between a white and a purple, the purple would have to be simplex"],"execution_count":7,"outputs":[{"output_type":"stream","text":["{'nulliplex': 18, 'simplex': 18, 'duplex': 0, 'triplex': 0, 'quadriplex': 0}\n"],"name":"stdout"}]},{"cell_type":"markdown","metadata":{"id":"QpUc-ZMq5kEd"},"source":["### Question 3\n","What proportion of purple-flowered progeny would be expected in the F2 generation if one of the purple-flowered F1 plants is self-pollinated? "]},{"cell_type":"code","metadata":{"id":"r27seEZ25oJS","colab":{"base_uri":"https://localhost:8080/","height":214},"executionInfo":{"status":"error","timestamp":1616452317701,"user_tz":360,"elapsed":588,"user":{"displayName":"Ammon Knaupp","photoUrl":"","userId":"01156806751079449669"}},"outputId":"45b8bf61-d2e7-47e0-9f27-b6c41b44890a"},"source":["purplef1 = Polyploidic_Assortment('C1','c1','c1','c1', dominant_allele='C1') # purple f1 progeny are C1 simplex\n","print(purple.cross(purplef1))\n","\n","# Answer: 18 simplex + 9 duplex -> 27 purple / 36 total = 0.75"],"execution_count":8,"outputs":[{"output_type":"error","ename":"NameError","evalue":"ignored","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)","\u001b[0;32m<ipython-input-8-ef89069d9bb1>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mpurplef1\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mPolyploidic_Assortment\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'C1'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'c1'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'c1'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'c1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdominant_allele\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'C1'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;31m# purple f1 progeny are C1 simplex\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpurple\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcross\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpurplef1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;31m# Answer: 18 simplex + 9 duplex -> 27 purple / 36 total = 0.75\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mNameError\u001b[0m: name 'purple' is not defined"]}]},{"cell_type":"markdown","metadata":{"id":"aZjJRXK-YTt3"},"source":["# Context for questions 4, 5, and 6\n","Drosophila fruit flies are diploids (2n = 2x = 8)."]},{"cell_type":"markdown","metadata":{"id":"16YMYkuEYfHf"},"source":["## Question 4\n","How many total chromosomes would a monosomic diploid fly (2n - 1) have in a somatic cell: "]},{"cell_type":"code","metadata":{"id":"jJBgXG3sYm6E","executionInfo":{"status":"aborted","timestamp":1616452317694,"user_tz":360,"elapsed":580,"user":{"displayName":"Ammon Knaupp","photoUrl":"","userId":"01156806751079449669"}}},"source":["# Assumptions\n","# monosomic means that one of its chromosomes is missing\n","\n","# Answer: 2n = 8 -> 8 - 1 = 7"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"ggU5QHP0Z_k9"},"source":["## Question 5\n","Suppose a monosomic chromosome IV female Drosophila fly that has white eyes (a recessive X-linked trait) and normal bristles is crossed with a male with a diploid set of chromosomes and normal red eye color but homozygous for the recessive chromosome IV bristle mutation shaven (sv).  In the F1 generation, what is the expected proportion of white-eyed, shaven-bristled progeny? "]},{"cell_type":"code","metadata":{"id":"shYTBvL8aGRS","executionInfo":{"status":"aborted","timestamp":1616452317695,"user_tz":360,"elapsed":580,"user":{"displayName":"Ammon Knaupp","photoUrl":"","userId":"01156806751079449669"}}},"source":["# Assumptions:\n","\n","# Female: \n","#    IV: S_\n","#     X: ww\n","# gametes: S,w ; _,w\n","\n","# Male:\n","#    IV: ss\n","#     X: WY\n","# gametes: s,W ; s,Y\n","\n","# F1 zygotes: Ss,Ww; Ss,wY; s_,Ww; s_,wY \n","\n","# Answer: proportion of ss(or s_), ww(or wY) (shaven-bristled, white-eyed progeny) is 1/4"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"6A7YUig2_xhv"},"source":["## Question 6\n","Eyeless is a recessive gene on chromosome 4 of Drosophila melanogaster. Flies homozygous for ey have tiny eyes or no eyes at all.  A male fly trisomic for chromosome 4 with genotype +/+/ey is crossed with a normal, diploid eyeless female. \n","\n","What proportion of their progeny will be eyeless? "]},{"cell_type":"code","metadata":{"id":"WHv3AI5N_6Iq","executionInfo":{"status":"aborted","timestamp":1616452317696,"user_tz":360,"elapsed":579,"user":{"displayName":"Ammon Knaupp","photoUrl":"","userId":"01156806751079449669"}}},"source":["# Female: ey/ey\n","# Male +/+/ey\n","\n","# male gametes  : +/+, ey, +/ey, +, +/ey, +\n","# female gametes: ey\n","# zygotes       : +/+/ey, ey/ey, 2(+/ey/ey), 2(+/ey),\n","\n","# Answer: proportion of ey/ey (eyeless) progeny is 1/6"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"bs1AnhBGBu-J"},"source":["# Context for questions 7, 8 and 9\n","Perennial desert shrubs of the genus Atriplex are typically dioecious; they consist of separate male and female plants.  Species within this genus also have varying levels of ploidy.  Four-wing saltbush (A. canescens) and salt sage (A. tridentata) are wide-spread species throughout the American west.  Both are fully fertile.  In 1979, Stutz et al. (Amer. J. Bot. 66:1181-1193) reported the discovery of an isolated female four-wing saltbush plant surrounded by a large population of salt sage near the south shore of the Great Salt Lake.  Chromosomes counts in the somatic cells of the salt sage population consistently yielded 2n = 54.  Seed collected from the isolated female four-wing saltbush plant were grown in a garden.  The plants that arose from the seeds were sterile and had 45 chromosomes in the somatic cells."]},{"cell_type":"markdown","metadata":{"id":"-sCviRbiB7ri"},"source":["## Question 7\n","How many chromosomes are in the somatic cells of the four-wing saltbush female plant?"]},{"cell_type":"code","metadata":{"id":"T02ZM_LaCA-t","executionInfo":{"status":"aborted","timestamp":1616452317697,"user_tz":360,"elapsed":579,"user":{"displayName":"Ammon Knaupp","photoUrl":"","userId":"01156806751079449669"}}},"source":["saltbush = Species(Monoploid_Genome(name='S'), total_chromosomes=36)\n","saltsage = Species(Monoploid_Genome(name='S'), total_chromosomes=54)\n","\n","# saltsage n = 27\n","# progeny n = 45\n","# therefore, saltbush n = 45 - 27 = 18\n","# therefore, saltbush total_chromosomes = 36\n","# Answer:                                ^^^"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"IpGXTb-BCCte"},"source":["## Question 8\n","What is the ploidy of the four-wing saltbush plant?"]},{"cell_type":"code","metadata":{"id":"GUpRIRFZCJof","executionInfo":{"status":"aborted","timestamp":1616452317698,"user_tz":360,"elapsed":577,"user":{"displayName":"Ammon Knaupp","photoUrl":"","userId":"01156806751079449669"}}},"source":["# Assumptions:\n","# x is the same for both plants\n","\n","# x must be a factor of the total number of chromosomes in both species as well as their hybrid\n","# x = 9 since 9 is the greatest common factor they all share (is this always true?).\n","\n","saltbush = Species(Monoploid_Genome(name='S'), x=9, total_chromosomes=36)\n","saltsage = Species(Monoploid_Genome(name='S'), x=9, total_chromosomes=54)\n","\n","print('Saltbush:', saltbush.toString())\n","print('Saltsage:', saltsage.toString())\n","\n","# Answer: saltbush ploidy is tetraploid"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"vDzxFuf6CJ-3"},"source":["## Question 9\n","What is the ploidy of the sterile progeny plants grown in the garden."]},{"cell_type":"code","metadata":{"id":"hB3fM663CPMg","executionInfo":{"status":"aborted","timestamp":1616452317698,"user_tz":360,"elapsed":574,"user":{"displayName":"Ammon Knaupp","photoUrl":"","userId":"01156806751079449669"}}},"source":["progeny = Species(Monoploid_Genome(name='H'), x=9, total_chromosomes=45)\n","print('Progeny:', progeny.toString())"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"YSsaWEBECPwF"},"source":["## Question 10\n","Mendelian ratios are modified in crosses involving autotetraploids.  In an autotetraploid pea species, green seed is governed by the dominant gene I (i.e., I--- plants have yellow seed, while iiii plants have green seed) and round seed shape is governed by the dominant gene R (i.e., R--- plants have round seed, while rrrr plants have wrinkled seed).   \n","\n","What is the predicted proportion of green round seed in the F2 progeny of a “dihybrid” cross involving the two independently assorting characteristics (e.g., IIIIRRRR X iiiirrrr)."]},{"cell_type":"code","metadata":{"id":"mwLJbocICWtL","executionInfo":{"status":"aborted","timestamp":1616452317699,"user_tz":360,"elapsed":574,"user":{"displayName":"Ammon Knaupp","photoUrl":"","userId":"01156806751079449669"}}},"source":["# quasi-dihybrid cross\n","\n","p1 = Polyploidic_Assortment(['I','I','I','I'],['R','R','R','R'])\n","p2 = Polyploidic_Assortment(['i','i','i','i'],['r','r','r','r'])\n","\n","f1 = Polyploidic_Assortment(['I','I','i','i'],['R','R','r','r'])\n","\n","# f1 gene 1 gametes: II ii 4*(Ii)\n","# f1 gene 2 gametes: RR rr 4*(Rr)\n","# smartest way of doing this: use independent assortment and product rule of probability.\n","# the probability of a particular combination of phenotypes (or genotypes) is equal to product the probabities of the individual events.\n","# the phenotype we want is green, round seeds. green seed is recessive, so nulliplex I. round seeds are dominant, so only one dominant allele is required.\n","# doing a punnett square considering only gene 1 shows that only 1 of the 36 products is nulliplex (iiii)\n","# we only need one punnett square b/c both sets of gametes are the same as far as zygosity\n","# therefore, the other 35 are non-nulliplex, meaning they have at least one dominant allele\n","\n","f2 = Polyploidic_Assortment([])\n","\n","# Answer: Proportion of green, round seed in the F2 progeny is 1/36 * 35/36 = 35/1296"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"gWfCkPDR4etF"},"source":["# Code logic tests"]},{"cell_type":"markdown","metadata":{"id":"5eI0h6Ix6evs"},"source":["## Zebroid test\n"]},{"cell_type":"code","metadata":{"id":"aUqC7W7LwFVE","executionInfo":{"status":"aborted","timestamp":1616452317699,"user_tz":360,"elapsed":572,"user":{"displayName":"Ammon Knaupp","photoUrl":"","userId":"01156806751079449669"}}},"source":["# zebroid test\n","\n","mountain_zebra = Species(Monoploid_Genome(name='M'), Monoploid_Genome(name='M'), total_chromosomes=32)\n","plains_zebra = Species(Monoploid_Genome(name='P'), Monoploid_Genome(name='P'), total_chromosomes=44)\n","grévys_zebra = Species(Monoploid_Genome(name='G'), Monoploid_Genome(name='G'), total_chromosomes=46)\n","kiang = Species(Monoploid_Genome(name='K'), Monoploid_Genome(name='K'), total_chromosomes=52)\n","onager = Species(Monoploid_Genome(name='O'), Monoploid_Genome(name='O'), total_chromosomes=56)\n","domesticated_donkey = Species(Monoploid_Genome(name='D'), Monoploid_Genome(name='D'), total_chromosomes=62)\n","horse = Species(Monoploid_Genome(name='H'), Monoploid_Genome(name='H'), total_chromosomes=64)\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"WEQ_xktz7F2M"},"source":["## uneven diploid number test"]},{"cell_type":"code","metadata":{"id":"33ea3jFlzfN3","executionInfo":{"status":"aborted","timestamp":1616452317700,"user_tz":360,"elapsed":572,"user":{"displayName":"Ammon Knaupp","photoUrl":"","userId":"01156806751079449669"}}},"source":["# certain numbers, when divided by 2, yield an odd number (6, 22, 34), and others yield even numbers (8, 16, 36)\n","\n","species1 = Species(Monoploid_Genome(name='A'), Monoploid_Genome(name='A'), total_chromosomes=32)   # even value of 'n'\n","species2 = Species(Monoploid_Genome(name='B'), Monoploid_Genome(name='B'), total_chromosomes=34) # odd value of 'n'\n","print(species1.toString())\n","print(species2.toString())\n","\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"ibHrNMgkIsS8"},"source":["## polyploidy tests"]},{"cell_type":"code","metadata":{"id":"ZRIKfieBIzRU","executionInfo":{"status":"aborted","timestamp":1616452317700,"user_tz":360,"elapsed":570,"user":{"displayName":"Ammon Knaupp","photoUrl":"","userId":"01156806751079449669"}}},"source":["gamete = Species(Monoploid_Genome(name='C', x=2)) # 2n = 2x = 4\n","print('Gamete:', gamete.toString(), gamete.gamete)\n","\n","diploid1 = Species(Monoploid_Genome(name='A', x=2), Monoploid_Genome(name='A', x=2)) # 2n = 2x = 4\n","diploid2 = Species(Monoploid_Genome(name='B', x=2), Monoploid_Genome(name='B', x=2)) # 2n = 2x = 4\n","print('Diploid1:', diploid1.toString(), diploid1.gamete, diploid1.unreduced_gamete)\n","\n","triploid = Species(Monoploid_Genome(name='B', x=2), Monoploid_Genome(name='B', x=2), Monoploid_Genome(name='A', x=2)) # 2n = 3x = 4\n","print('Triploid1:', triploid.toString(), triploid.gamete)\n","\n","# allotetraploid (allo- is a prefix which means \"different\" or \"other\")\n","# allotetraploid = diploid1.cross(diploid2)\n","\n","# autotetraploid (auto- is a prefix which means \"self\" or \"same\")\n","# allotetraploid = diploid1.cross(diploid1)\n","\n","# Bread wheat (AABBDD)\n","# 2n = 6x = 42 (n=21, x=7)\n","# Gamete is ABD\n","\n","# Durum wheat (AABB)\n","# 2n = 4x = 28\n","# Gamete is AB\n","\n","# cross the two and get pentaploid (AABBD)\n","# sterile due to inablity to synapse"],"execution_count":null,"outputs":[]}]}